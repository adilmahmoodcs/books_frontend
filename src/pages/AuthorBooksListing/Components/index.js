import { AuthorBooksHeader } from "./AuthorBooksHeader";

export { AuthorBooksHeader };
