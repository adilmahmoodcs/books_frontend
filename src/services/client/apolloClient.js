import { ApolloClient, InMemoryCache, HttpLink } from '@apollo/client'
import fetch from 'cross-fetch'

export const client = new ApolloClient({
  link: new HttpLink({
    uri: `${import.meta.env.VITE_API_URL}/graphql`,
    fetch
  }),
  cache: new InMemoryCache()
})
