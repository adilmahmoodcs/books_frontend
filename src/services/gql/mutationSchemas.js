import { gql } from "@apollo/client";

export const CREATE_AUTHOR = gql`
  mutation CreateAuthor($author: AuthorCreateInput!) {
    authorCreate(input: $author) {
      author {
        id
        firstname
        surname
        dob
      }
    }
  }
`;

export const DELETE_AUTHOR = gql`
  mutation DeleteAuthor($authorId: AuthorDeleteInput!) {
    authorDelete(input: $authorId) {
      author {
        id
      }
    }
  }
`;

export const CREATE_BOOK = gql`
  mutation CreateBook($book: BookCreateInput!) {
    bookCreate(input: $book) {
      book {
        id
        title
      }
    }
  }
`;

export const DELETE_BOOK = gql`
  mutation DeleteBook($bookId: BookDeleteInput!) {
    bookDelete(input: $bookId) {
      book {
        id
      }
    }
  }
`;
