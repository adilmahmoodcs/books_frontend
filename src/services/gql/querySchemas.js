import { gql } from "@apollo/client";

export const GET_AUTHORS = gql`
  query GetAuthors {
    authors {
      id
      firstname
      surname
      dob
    }
  }
`;

export const GET_AUTHOR_BY_ID = (authorId) => gql`
  query GetAuthorById {
    author(id: ${authorId}) {
      firstname
      surname
      dob
      books {
        id
        title
        genre
        publishYear
      }
    }
  }
`;
