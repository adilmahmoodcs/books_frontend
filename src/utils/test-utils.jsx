/* eslint-disable import/export */
import { ApolloProvider } from "@apollo/client";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import client from "../services/client/apolloClient";

const customRender = (ui, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      // <ApolloProvider client={client}>
        <BrowserRouter>{children}</BrowserRouter>
      // </ApolloProvider>
    ),

    ...options,
  });

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";

// override render export
export { customRender as render };
