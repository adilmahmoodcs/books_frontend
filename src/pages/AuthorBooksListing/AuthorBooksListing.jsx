import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { useParams } from "react-router-dom";
import { useQuery } from "@apollo/client";
import { GET_AUTHOR_BY_ID } from "../../services/gql/querySchemas";
import AuthorBooksHeader from "./Components/AuthorBooksHeader";
import { Grid } from "@mui/material";
import { BookCard } from "../../components";
import { AddBook } from "../../components";

function AuthorBooksListing(props) {
  const { authorId } = useParams();
  const { data, error, loading } = useQuery(GET_AUTHOR_BY_ID(authorId));

  const [genres, setGenres] = useState([]);
  const [open, setOpen] = useState(false);

  useEffect(() => {
    if (!!data) {
      const _genres = [...new Set(data.author.books.map((book) => book.genre))];
      setGenres(_genres);
    }
  }, [data]);

  if (loading) return null;

  if (error) return null;

  return (
    <div>
      <AuthorBooksHeader
        data={data}
        genres={genres}
        onAddNewBookClick={() => setOpen(true)}
      />

      <Grid container spacing={3} sx={{ mt: 2 }}>
        {data.author.books.map((book) => (
          <Grid key={book.id} item lg={3}>
            <BookCard data={book} />
          </Grid>
        ))}
      </Grid>

      <AddBook
        open={open}
        handleClose={() => setOpen(false)}
        authorId={authorId}
      />
    </div>
  );
}

AuthorBooksListing.propTypes = {};

export default AuthorBooksListing;
