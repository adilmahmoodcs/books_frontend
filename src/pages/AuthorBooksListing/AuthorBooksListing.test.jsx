import { MockedProvider } from "@apollo/client/testing";
import { describe, expect, it } from "vitest";
import { GET_AUTHOR_BY_ID } from "../../services/gql/querySchemas";
import { render, waitFor } from "../../utils/test-utils";
import AuthorBooksListing from "./AuthorBooksListing";

const mocks = [
  {
    request: {
      query: GET_AUTHOR_BY_ID("1"),
    },
    result: {
      data: {
        author: [
          {
            dob: "2022-03-09T19:00:00Z",
            firstname: "Jane",
            surname: "Smith",
            books: [
              {
                genre: "Comedy",
                id: "1",
                publishYear: "2022-03-18T19:00:00Z",
                title: "On the road",
                __typename: "Book",
              },
            ],
            __typename: "Author",
          },
        ],
      },
    },
  },
];

const WrappedComponent = () => (
  <MockedProvider mocks={mocks} addTypename={false}>
    <AuthorBooksListing />
  </MockedProvider>
);

describe("it tests Author listing", () => {
  it("query is fetched", async () => {
    const { container } = render(<WrappedComponent />);
    await waitFor(() => new Promise((res) => setTimeout(res, 0)));
    expect(container).toMatchSnapshot();
  });
});
