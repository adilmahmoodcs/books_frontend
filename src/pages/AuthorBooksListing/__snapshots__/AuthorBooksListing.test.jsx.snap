// Vitest Snapshot v1

exports[`it tests Author listing > query is fetched 1`] = `<div />`;
